File: Difficulty: Medium/Longest Common Substring/longest-common-substring.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
  

  
    int longestCommonSubstr(string str1, string str2) {
        // your code here
        
        // don,t write code for shortest common supersequence
        int n = str1.length();
        int m = str2.length();
        
        // Dont use recursion******************************
        
        // use tabulation
        
        vector<vector<int>> dp(n+1,vector<int>(m+1,0));

        for(int i = 0; i <= n; i++)  dp[i][0] = 0;
        for(int j = 0; j <= m; j++)  dp[0][j] = 0;
        int ans = 0;

        for(int i = 1; i <= n; i++)
        {
            for(int j = 1; j <= m; j++)
            {
                if(str1[i-1] == str2[j-1]) 
                {
                    
                dp[i][j] = 1 + dp[i-1][j-1];
                ans = max(ans,dp[i][j]);
                }
        
                else
                dp[i][j] = 0;
            }
        }

        return ans;
        
        // vector<int> prev(m+1,0);
        // vector<int> curr(m+1,0);
        // int ans = 0;
        
        // for(int i = 1; i <= n; i++)
        // {
        //     for(int j = 1; j <= m; j++)
        //     {
        //         if(str1[i-1] == str2[j-1])
        //         {
        //         curr[j] = 1 + prev[j-1];
        //         ans = max(ans,curr[j]);
        //         }
        //         else
        //         curr[j] = 0;
        //     }
        //     prev = curr;
        // }
        
        // return ans;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string s1, s2;
        cin >> s1 >> s2;
        Solution ob;

        cout << ob.longestCommonSubstr(s1, s2) << endl;
    }
}
// } Driver Code Ends

File: Difficulty: Medium/Longest Common Substring/longest-common-substring.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            String S1 = read.readLine().trim();
            String S2 = read.readLine().trim();

            Solution ob = new Solution();
            System.out.println(ob.longestCommonSubstr(S1, S2));

            System.out.println("~");
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    public int longestCommonSubstr(String s1, String s2) {
        // code here
        int n = s1.length();
        int m = s2.length();
        int maxLen = 0;
        
        int[][] dp = new int[n+1][m+1];
        for(int i = 0; i <= n; i++)
        {
            Arrays.fill(dp[i],0);
        }
        
        for(int j = 0; j <= m; j++)
        {
            dp[0][j] = 0;
        }
        
        for(int i = 0; i <= n; i++)
        {
            dp[i][0] = 0;
        }
        
        for(int i = 1; i <= n; i++)
        {
            for(int j = 1; j <= m; j++)
            {
                if(s1.charAt(i-1) == s2.charAt(j-1))
                {
                    dp[i][j] = 1 + dp[i-1][j-1];
                    maxLen = Math.max(maxLen,dp[i][j]);
                }
                else
                {
                    dp[i][j] = 0;
                }
            }
        }
        
        return maxLen;
    }
}